File: artifacts/python-backend/app/services/indicators.py
"""
Technical indicator calculations using the `ta` library (pandas-based).
Falls back to manual numpy implementations if needed.
"""
import pandas as pd
import ta
from ta.trend import EMAIndicator, SMAIndicator, MACD
from ta.momentum import RSIIndicator
from ta.volatility import BollingerBands, AverageTrueRange


def _closes_series(data: list[float]) -> pd.Series:
    return pd.Series(data, dtype=float)


def _ohlcv_df(ohlcv: list[dict]) -> pd.DataFrame:
    return pd.DataFrame(ohlcv)


def calculate_ema(data: list[float], period: int) -> list[float]:
    if len(data) < period:
        return []
    s = _closes_series(data)
    ema_ind = EMAIndicator(close=s, window=period, fillna=False)
    result = ema_ind.ema_indicator().dropna().tolist()
    return result


def calculate_sma(data: list[float], period: int) -> list[float]:
    if len(data) < period:
        return []
    s = _closes_series(data)
    sma_ind = SMAIndicator(close=s, window=period, fillna=False)
    result = sma_ind.sma_indicator().dropna().tolist()
    return result


def calculate_rsi(data: list[float], period: int = 14) -> list[float]:
    if len(data) < period + 1:
        return []
    s = _closes_series(data)
    rsi_ind = RSIIndicator(close=s, window=period, fillna=False)
    result = rsi_ind.rsi().dropna().tolist()
    return result


def calculate_macd(data: list[float], fast: int = 12, slow: int = 26, signal: int = 9) -> dict:
    s = _closes_series(data)
    macd_ind = MACD(close=s, window_slow=slow, window_fast=fast, window_sign=signal, fillna=False)
    macd_line   = macd_ind.macd().dropna().tolist()
    signal_line = macd_ind.macd_signal().dropna().tolist()
    histogram   = macd_ind.macd_diff().dropna().tolist()
    return {"macd": macd_line, "signal": signal_line, "histogram": histogram}


def calculate_bollinger_bands(data: list[float], period: int = 20, sd: float = 2.0) -> dict:
    s = _closes_series(data)
    bb = BollingerBands(close=s, window=period, window_dev=sd, fillna=False)
    upper  = bb.bollinger_hband().dropna().tolist()
    middle = bb.bollinger_mavg().dropna().tolist()
    lower  = bb.bollinger_lband().dropna().tolist()
    return {"upper": upper, "middle": middle, "lower": lower}


def calculate_atr(ohlcv: list[dict], period: int = 14) -> list[float]:
    if len(ohlcv) < period + 1:
        return []
    df = _ohlcv_df(ohlcv)
    atr_ind = AverageTrueRange(
        high=df["high"].astype(float),
        low=df["low"].astype(float),
        close=df["close"].astype(float),
        window=period,
        fillna=False,
    )
    return atr_ind.average_true_range().dropna().tolist()


def calculate_vwap(ohlcv: list[dict]) -> list[float]:
    vwap = []
    cum_tp_vol = 0.0
    cum_vol = 0.0
    for d in ohlcv:
        tp = (d["high"] + d["low"] + d["close"]) / 3
        vol = d["volume"] or 0
        cum_tp_vol += tp * vol
        cum_vol += vol
        vwap.append(cum_tp_vol / cum_vol if cum_vol > 0 else tp)
    return vwap


def detect_sr(ohlcv: list[dict], lookback: int = 10) -> dict:
    highs = [d["high"] for d in ohlcv]
    lows = [d["low"] for d in ohlcv]
    supports: list[float] = []
    resistances: list[float] = []
    for i in range(lookback, len(ohlcv) - lookback):
        seg_h = highs[i - lookback:i + lookback + 1]
        seg_l = lows[i - lookback:i + lookback + 1]
        if highs[i] == max(seg_h):
            resistances.append(highs[i])
        if lows[i] == min(seg_l):
            supports.append(lows[i])
    return {
        "supports": sorted(set(supports)),
        "resistances": sorted(set(resistances)),
    }

File: artifacts/api-server/src/services/indicators.ts
export interface OHLCV { date: string; open: number; high: number; low: number; close: number; volume: number; }

export function calculateEMA(data: number[], period: number): number[] {
  if (data.length < period) return [];
  const k = 2 / (period + 1);
  const ema: number[] = [data.slice(0, period).reduce((a, b) => a + b, 0) / period];
  for (let i = period; i < data.length; i++) ema.push(data[i] * k + ema[ema.length - 1] * (1 - k));
  return ema;
}

export function calculateSMA(data: number[], period: number): number[] {
  const sma: number[] = [];
  for (let i = period - 1; i < data.length; i++) {
    sma.push(data.slice(i - period + 1, i + 1).reduce((a, b) => a + b, 0) / period);
  }
  return sma;
}

export function calculateRSI(data: number[], period = 14): number[] {
  if (data.length < period + 1) return [];
  const gains: number[] = [], losses: number[] = [];
  for (let i = 1; i < data.length; i++) {
    const c = data[i] - data[i - 1];
    gains.push(c > 0 ? c : 0);
    losses.push(c < 0 ? Math.abs(c) : 0);
  }
  let ag = gains.slice(0, period).reduce((a, b) => a + b, 0) / period;
  let al = losses.slice(0, period).reduce((a, b) => a + b, 0) / period;
  const rsi: number[] = [];
  for (let i = period; i < gains.length; i++) {
    ag = (ag * (period - 1) + gains[i]) / period;
    al = (al * (period - 1) + losses[i]) / period;
    rsi.push(100 - 100 / (1 + (al === 0 ? 100 : ag / al)));
  }
  return rsi;
}

export function calculateMACD(data: number[], fast = 12, slow = 26, signal = 9) {
  const ef = calculateEMA(data, fast), es = calculateEMA(data, slow);
  const diff = slow - fast;
  const macd = es.map((v, i) => ef[i + diff] - v);
  const sl = calculateEMA(macd, signal);
  return { macd, signal: sl, histogram: sl.map((v, i) => macd[i + signal - 1] - v) };
}

export function calculateBollingerBands(data: number[], period = 20, sd = 2) {
  const middle = calculateSMA(data, period);
  const upper: number[] = [], lower: number[] = [];
  middle.forEach((sma, idx) => {
    const s = data.slice(idx, idx + period);
    const variance = s.reduce((sum, v) => sum + Math.pow(v - sma, 2), 0) / period;
    const std = Math.sqrt(variance);
    upper.push(sma + sd * std);
    lower.push(sma - sd * std);
  });
  return { upper, middle, lower };
}

export function calculateATR(ohlcv: OHLCV[], period = 14): number[] {
  if (ohlcv.length < 2) return [];
  const trs: number[] = [];
  for (let i = 1; i < ohlcv.length; i++) {
    const h = ohlcv[i].high, l = ohlcv[i].low, pc = ohlcv[i - 1].close;
    trs.push(Math.max(h - l, Math.abs(h - pc), Math.abs(l - pc)));
  }
  return calculateSMA(trs, period);
}

export function detectSR(ohlcv: OHLCV[], lookback = 10) {
  const highs = ohlcv.map(d => d.high), lows = ohlcv.map(d => d.low);
  const supports: number[] = [], resistances: number[] = [];
  for (let i = lookback; i < ohlcv.length - lookback; i++) {
    const lh = highs.slice(i - lookback, i + lookback + 1);
    const ll = lows.slice(i - lookback, i + lookback + 1);
    if (highs[i] === Math.max(...lh)) resistances.push(highs[i]);
    if (lows[i] === Math.min(...ll)) supports.push(lows[i]);
  }
  return { supports: [...new Set(supports)].sort((a, b) => a - b), resistances: [...new Set(resistances)].sort((a, b) => a - b) };
}
